File: backend/src/main/java/com/hpms/backend/model/Building.java
package com.hpms.backend.model;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.time.LocalDateTime;
import java.util.Collection;
import java.util.HashSet;

@Entity
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Table(name = "buildings")
public class Building {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false)
    private String name;

    @Column(columnDefinition = "TEXT")
    private String description;

    private String address;
    private String city;
    private String zipcode;
    private String country;
    private String phoneNumber;
    private String email;

    private LocalDateTime createdAt;
    private LocalDateTime updatedAt;
    private Boolean active = true;

    @ManyToMany(fetch = FetchType.EAGER, cascade = {
            CascadeType.DETACH, CascadeType.MERGE, CascadeType.PERSIST, CascadeType.REFRESH
    })
    @JoinTable(name = "user_buildings",
            joinColumns = @JoinColumn(name = "building_id", referencedColumnName = "id"),
            inverseJoinColumns = @JoinColumn(name = "user_id", referencedColumnName = "id"))
    private Collection<User> users = new HashSet<>();

    @OneToMany(mappedBy = "building", fetch = FetchType.LAZY)
    private Collection<Room> rooms = new HashSet<>();

    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
        updatedAt = LocalDateTime.now();
    }

    @PreUpdate
    protected void onUpdate() {
        updatedAt = LocalDateTime.now();
    }
}

File: frontend/src/interfaces/building.ts
import type { FormValues } from "../lib/form";

export interface BuildingData {
	id: number;
	name: string;
	address: string;
	city: string;
	zipcode: string;
	description?: string;
	country: string;
	phoneNumber: string;
	email: string;
	createdAt: Date;
	updatedAt?: Date;
	active?: boolean;
	userIds?: number[];
	roomIds?: number[];
}

export interface BuildingFormData extends FormValues {
	name: string;
	address: string;
	city: string;
	zipcode: string;
	country: string;
	phoneNumber: string;
	email: string;
	description: string;
	active: boolean;
}
